src/FOMObot/Helpers/Algorithm.hs

```diff
@@ -1,44 +1,56 @@
 module FOMObot.Helpers.Algorithm
-    ( shiftInTime
+    ( shiftInHistory
     , shiftInEvent
-    , calcDensity
     , detectFOMOEvent
     ) where
 
-import Control.Lens ((^.))
+import Control.Lens ((^.), (^?), (^?!), (&), (.~), (%~), _head, _last, views)
+import Data.List (nub)
 import qualified Web.Slack as Slack
 
 import FOMObot.Types.Bot
 import FOMObot.Types.BotConfig
 import FOMObot.Types.ChannelState
+import FOMObot.Types.HistoryItem
 
 type Density = Double
 
 calcDensity :: ChannelState -> Bot Density
-calcDensity ChannelState{stateHistory} = do
+calcDensity s = do
     BotConfig{configHistorySize} <- getConfig
-    return $ if isArrayFull stateHistory configHistorySize
+    return $ if isArrayFull (s ^. stateHistory) configHistorySize
         then calc $ fromIntegral configHistorySize
         else 0
   where
     calc historySize = 60 * historySize / timeOverHistory
-    timeOverHistory = realToFrac $ (head stateHistory ^. Slack.slackTime) - (last stateHistory ^. Slack.slackTime)
-
-detectFOMOEvent :: Density -> Bot Bool
-detectFOMOEvent density = do
-    BotConfig{configThreshold} <- getConfig
-    return $ density > configThreshold
+    timeOverHistory = realToFrac $ latestTimeStamp - earliestTimeStamp
+    latestTimeStamp = s ^?! stateHistory . _head . historyTimeStamp . Slack.slackTime
+    earliestTimeStamp  = s ^?! stateHistory . _last . historyTimeStamp . Slack.slackTime
+
+detectFOMOEvent :: ChannelState -> Bot Bool
+detectFOMOEvent state = do
+    densitySurpassesThreshold <- (>) <$> calcDensity state <*> (configThreshold <$> getConfig)
+    return $ and
+        [ densitySurpassesThreshold
+        , atLeastThreeUniqueUsers state
+        ]
+  where
+    atLeastThreeUniqueUsers = views stateHistory ((>=3) . length . nub . (map (^. historyUserId)))
 
-shiftInTime :: BotConfig -> Slack.SlackTimeStamp -> ChannelState -> ChannelState
-shiftInTime BotConfig{configHistorySize} timestamp s@ChannelState{stateHistory} =
-    s { stateHistory = shiftIn configHistorySize stateHistory timestamp }
+shiftInHistory :: BotConfig -> HistoryItem -> ChannelState -> ChannelState
+shiftInHistory BotConfig{configHistorySize} historyItem s =
+    if (s ^? stateHistory . _head . historyUserId) == Just (historyItem ^. historyUserId)
```

Comment:
I should have access to `s` and `historyItem` in the `where` so maybe a function like `isFromPreviousUser`? other suggestions?